File: leetcode/0260-single-number-iii/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int xorAll = 0;
        for (int n : nums) {
            xorAll ^= n;
        }

        int diffBit = xorAll & (-xorAll);

        int a = 0;
        for (int n : nums) {
            if (n & diffBit) {
                a ^= n;
            }
        }

        int b = xorAll ^ a;
        return {a, b};
    }
};

File: leetcode/0260-single-number-iii/solution.py
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        xor_all = 0
        for n in nums:
            xor_all ^= n

        # isolate the lowest set bit
        diff_bit = xor_all & (-xor_all)

        a = 0
        for n in nums:
            if n & diff_bit:
                a ^= n

        b = xor_all ^ a
        return [a, b]
